### PR TITLE
Add TM stats Per User

### LINKED ISSUE #115: Endpoint  Returning result in negative numbers
```

{
    "fromTimestamp": "2021-12-28T07:00:00.330",
    "toTimestamp": "2021-12-28T15:22:21.330",
    "projectIds": [],
    "hashtags": [
        "maproulette"
    ]
}
```
With the help of @d-rita  While passing this parameter to detailed report 
I can see endpoint is returning this type of result for mapped tasks, validated tasks with negative number , Is it expected output ? Why we are even accepting negative numbers ? It should be positive integer instead , Is there any case for negative scenario ?
```
    {
      "userId": ###,
      "username": "###",
      "totalBuildings": 6,
      "mappedTasks": -1,
      "validatedTasks": -1,
      "editors": "iD 2.20.2,"
    }
```
cc: @ramyaragupathy  @omranlm

### LINKED ISSUE #147: Add time spent mapping and validation to the mapathon detailed endpoint

### LINKED ISSUE #156: Add unit tests for TM queries for time spent mapping and validating per user

### LINKED ISSUE #158: Modify existing tests for user contributor query
The user contributor query has been updated and no longer queries the tasks mapped and tasks validated per user via a TM dblink. This is now separated into alternate TM queries. Hence the test has to be updated to match the query.

### PR DESCRIPTION
**Pull request by d-rita**
**What does this PR do?**
- This PR adds stats for the `time spent mapping` and `time spent validating` per user within a given time period under the mapathon detail endpoint. 
- It also replaces the dblink functions extracting stats for the `tasks_mapped_per_user` and `tasks_validated_per_user` with direct queries from the Tasking Manager database. 
- It validates against negative tasks mapped and validated counts.

**How to test:** 
- Connect to both the Insights and TM databases. 
- Test with inputs for the mapathon detail endpoint, i.e. start date, end date, project ids, and/or hashatags via `/mapathon/detail`
- Expected response should be an object with `mappedFeatures`, `contributors`, and the `tmStats` keys. The `tmStats` contains `tasks_mapped`, `tasks_validated`, `time_spent_mapping (s)`, and `time_spent_validating (s)`stats per user for the given projects or hashtag in the given time range (that is < 24 hours).

**Example Queries**: 
1. query with project ids: `{
    "fromTimestamp": "2021-02-21T18:00:00.000",
    "toTimestamp": "2021-02-22T09:00:00.000",
    "projectIds": [
        10071,
        10334,
        9718,
        10191
    ],
    "hashtags": []
}`
2. query with TM specific hashtag: `{
    "fromTimestamp": "2021-02-21T18:00:00.000",
    "toTimestamp": "2021-02-22T09:00:00.000",
    "projectIds": [],
    "hashtags": [
        "hotosm-project-10334"
    ]
}`
3. query with non-TM specific hashtag: `{
    "fromTimestamp": "2021-02-21T18:00:00.000",
    "toTimestamp": "2021-02-22T09:00:00.000",
    "projectIds": [],
    "hashtags": [
        "missingmaps"
    ]
}`
4. query with non-existent hashtag: `{
    "fromTimestamp": "2021-02-21T18:00:00.000",
    "toTimestamp": "2021-02-22T09:00:00.000",
    "projectIds": [],
    "hashtags": [
        "artemis"
    ]
}`

**Related issues**
- Fixes #147 
- Fixes #158
- Fixes #156
- Fixes #115

### CODE REVIEW

File: tests/src/test_app.py [hunk 1]
```diff
@@ -251,6 +251,22 @@ def test_mapathon_users_contributors_mapathon_query_builder():
...
+    params = mapathon_validation.MapathonRequestParams(**test_param)
+    result_time_spent_mapping_and_validating_query = mapathon_query_builder.create_user_time_spent_mapping_and_validating_query(params)
+    print(result_time_spent_mapping_and_validating_query.encode('utf-8'))
```
Comment (itskshitiz321) [inline tests/src/test_app.py:266]: Avoid Printing test queries !

File: tests/src/test_app.py [hunk 2]
```diff
@@ -251,6 +251,22 @@ def test_mapathon_users_contributors_mapathon_query_builder():
...
+    params = mapathon_validation.MapathonRequestParams(**test_param)
+    result_tasks_mapped__and_validated_query = mapathon_query_builder.create_user_tasks_mapped_and_validated_query(params)
+    print(result_tasks_mapped__and_validated_query.encode('utf-8'))
```
Comment (itskshitiz321) [inline tests/src/test_app.py:258]: Same here as well

**Review by kshitijrajsharma — COMMENTED** [review]
Code Wise all good from my side ! Will test it with endpoint itself after I get credentials for TM  and leave comment if found any

**Comment by itskshitiz321** [comment]
It is expected that this PR should resolve this issue as well 

- Resolve #115

File: src/galaxy/query_builder/builder.py [hunk 3]
```diff
@@ -204,23 +204,68 @@ def create_users_contributions_query(params, changeset_query):
...
-            '{from_timestamp}',
-            '{to_timestamp}') AS editors
+        '{from_timestamp}',
+        '{to_timestamp}') AS editors
     FROM T3;
     """
     return query
 
+def create_user_time_spent_mapping_and_validating_query(params):
+    project_ids = ",".join([str(p) for p in params.project_ids])
```
Comment (itskshitiz321) [inline src/galaxy/query_builder/builder.py:215]: We have several conditions for Mapathon Reports , For eg: Project ID can be empty it is optional field user may can only supply hashtag or only project id , This generator will work for those cases where project id is supplied but will not work when hashtag is supplied , Here is the sample query output of this function 
```
        SELECT DISTINCT th.user_id,
            (
                SELECT COUNT(task_id)
                FROM public.task_history
                WHERE user_id = th.user_id
                AND action_text = 'MAPPED'
                AND action_date BETWEEN '2021-09-23 18:30:00' AND '2021-09-24 18:15:00'
                AND project_id IN ()
            ) AS mapped_task_count,
            (
                SELECT COUNT(task_id)
                FROM public.task_history
                WHERE user_id = th.user_id
                AND action_text = 'VALIDATED'
                AND action_date BETWEEN '2021-09-23 18:30:00' AND '2021-09-24 18:15:00'
                AND project_id IN ()
            ) AS validated_task_count
        FROM public.task_history th
        WHERE th.action_text = 'MAPPED' OR th.action_text = 'VALIDATED'
```

File: src/galaxy/query_builder/builder.py [hunk 4]
```diff
@@ -204,23 +204,68 @@ def create_users_contributions_query(params, changeset_query):
...
+                AND project_id IN ({project_ids})
+            ) AS time_spent_validating
+        FROM public.task_history th
+        WHERE (action = 'LOCKED_FOR_MAPPING' or action = 'AUTO_UNLOCKED_FOR_MAPPING') or action = 'LOCKED_FOR_VALIDATION'
+        GROUP BY th.user_id;
+    """
+    
+    return query;
+
+def create_user_tasks_mapped_and_validated_query(params):
```
Comment (itskshitiz321) [inline src/galaxy/query_builder/builder.py:243]: Similar condition here as well , new condition is needed for hashtags !

Comment (itskshitiz321) [reply src/galaxy/query_builder/builder.py:243]: You can Checkout validation models for input fields !

**Review by kshitijrajsharma — CHANGES_REQUESTED** [review]
API breakdowns with multiple combination of input supported for Mapathon detailed report !

**Review by omranlm — APPROVED** [review]
Reviewed the queries to get tasks mapped and validated and they match what we had in the insights before

**Review by kshitijrajsharma — APPROVED** [review]

**Pull request merged**